feat(SAPIC-257): implementation of functionality for proxying commands channels and streams

brutusyhy opened the pull request:
Implement a `tauri-plugin-dapis` that will host an http/websocket server for proxying tauri commands, channels and streams. 
The `examples/test-app` provides following example endpoints:
POST "/greet": Responds with a welcome message based on the input
POST "/set-state": Set a tauri-managed string state to a different value, and returns the old state
websocket "/stream": Given an upper limit, stream the integers from 0 up to that number every 100ms.
POST "/emit": Emit an event with a given event name to all targets

In addition, the plugin can be set up to listen to specific tauri events. It will stream those events on websocket endpoint "/{event_name}"

By default, the server runs on 127.0.0.1:12345. The input should be passed as JSON in the request body.

brutusyhy committed:
example of http/websocket server for proxying tauri commands

brutusyhy committed:
extract server logic into `tauri-plugin-dapis`

brutusyhy committed:
use `tauri-plugin-dapis` in the `dapis-test` demo

brutusyhy committed:
reorganize project structure

Move plugin code to src/ and test app to examples/test-app/

g10z3r commented on src/lib.rs:
```diff
@@ -0,0 +1,93 @@
+}
+
+pub struct Builder<R: Runtime> {
+    routes: HashMap<String, MethodRouter<AxumState<R>>>,
```
Maybe it makes sense to represent all the route as an enum?

```rust
enum Endpoint {
    Route{...},
    Channel{...},
    Stream{...},
}
```

brutusyhy replied:
Right now there's no distinction between channel and stream for us: Tauri channel correspond to a websocket stream, but I think having such a enum might make sense.

g10z3r commented on src/lib.rs:
```diff
@@ -0,0 +1,93 @@
+use tauri::{AppHandle, Runtime};
+const DEFAULT_PORT: usize = 12345;
+
+pub struct AxumState<R: Runtime> {
```
Is this the state of the library or the plugin?

brutusyhy replied:
It's held by the axum router so that it has access to app state.

g10z3r commented on src/lib.rs:
```diff
@@ -0,0 +1,93 @@
+                let axum_router = axum_router.with_state(state);
+
+                // We use a channel to propagate error out of this future
+                let (tx, rx) = channel();
```
If we give it a more descriptive name, we won’t need to write a comment explaining what the channels are 😄 for example, `errors_tx`.

brutusyhy committed:
rename channel name to be more descriptive

brutusyhy committed:
implement `emit` and `listen` endpoints for proxying tauri events

brutusyhy committed:
implement event listener setup for tauri-plugin-dapis

Allow specifying certain events that should be listened. The server will stream the corresponding events on /{event_name} websocket endpoints.

brutusyhy committed:
remove old listen handler

brutusyhy committed:
Create README
